Python/chapter03/p06_animal_shelter/miguelHx.py

```diff
@@ -0,0 +1,222 @@
+"""3.6 - Animal Shelter:
+An animal shelter, which holds only dogs and cats, operates on a strictly
+"first in, first out" basis. People must adopt either the "oldest" (based on arrival time)
+animal of all animals at the shelter, or they can select whether they would prefer
+a dog or a cat (and will receive the oldest animal of that type). They cannot
+select which specific animal they would like. Create the data structures to maintain this system
+and implement operations such as enqueue, dequeueAny, dequeueDog, and
+dequeueCat. You may use the built-in LinkedList data structure.
+"""
+import unittest
+
+from collections import deque
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+
+
+class AnimalTypeEnum(Enum):
+    DOG = 'dog'
+    CAT = 'cat'
+
+
+@dataclass
+class Animal:
+    name: str
+    species: AnimalTypeEnum
+    timestamp: Optional[int] = None
```

Comment:
from the typing perspective, the timestamp is not an inherent property of the animal, and there is nothing preventing a caller from fiddling with the timestamp from the outside, confusing the implementation. furthermore, it is not possible to express that an animal that currently exists in the shelter is guaranteed to have a timestamp purely from the type system.

there _is_ a way to solve both of these problems! by creating a wrapper class

```python
@dataclass
class QueuedAnimal:
    animal: Animal
    timestamp: int
```

and wrapping `Animal`s into `QueuedAnimal`s upon enqueueing.

Reply:
Thank you !  Updated here:  ffb663aad36f9395ae08912bdc2513bfb2832d11

How does one get really good at designing classes?

Reply:
There are two ideas that can give you very good bang for the buck:

* Make interfaces hard to misuse: try to constrain the options available to callers to avoid them (accidentally) falling into mistakes. This can be achieved by providing higher-level methods (instead of requiring the caller to chain several lower-level methods in sequence), removing unnecessary fields from the public interface, creating factory functions that return fully-initialized objects (instead of relying on docstrings/examples that tell the caller how to initialize it), bundling sets of required parameters into classes or structures to make it obvious that they _need_ to be all defined together, etc. There's a good little article here: https://testing.googleblog.com/2018/07/code-health-make-interfaces-hard-to.html
* There's an idiom in functional programming called "make illegal states unrepresentable", which means that folks should design their data structures in a way that invalid combinations of fields in a class cannot even be expressed. There's a good article about it here, taking it further and applying this idea in TypeScript through Domain-Driven Design: https://khalilstemmler.com/articles/typescript-domain-driven-design/make-illegal-states-unrepresentable/

Reply:
oh right, i forgot: there's also discouraging the use of `Optional`: inside a class, every `Optional` makes the number of representable states to double! odds are, some of those 2^n combinations aren't valid. by moving the fields marked `Optional` to another class (like we did in this thread), or bundling them all into a class to signify that they must all be provided or are all optional _together_ (say, there's a class that allows a different functionality for anonymous users and logged in users that have completed their profile, the set of (session_id, user_id, profile) can be bundled into a class that _itself_ can be marked as `Optional` altogether).